REGRESSION (Safari 26.2): shape-outside initially incorrect when used with imported font

nipunshukla21 opened the pull request:
#### acbb891ea04ca90cb32b107cfb87e9ae1f3214f4
<pre>
REGRESSION (Safari 26.2): shape-outside initially incorrect when used with imported font
<a href="https://bugs.webkit.org/show_bug.cgi?id=304019">https://bugs.webkit.org/show_bug.cgi?id=304019</a>
<a href="https://rdar.apple.com/166336491">rdar://166336491</a>

Reviewed by Simon Fraser and Alan Baradlay.

When web fonts load after initial layout, CSS shape properties remain unchanged
but box dimensions can change due to different font metrics. The regression caused
an early return when CSS is identical, preventing shape invalidation.

This issue was fixed by extending the early return condition to check whether
Style::Difference requires a repaint or layout. This ensures shapes are
invalidated when dimension-affecting properties change.

* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-invalidation-after-font-load-expected.txt: Added.
* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-invalidation-after-font-load.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):
(WebCore::RenderBox::updateShapeOutsideInfoAfterStyleChange):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: <a href="https://commits.webkit.org/305299@main">https://commits.webkit.org/305299@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9753f2a74a15c8aa0d74752de9ce0f83a700c940

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137960 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146028 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90935 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11026 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10468 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105501 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/76994 "Build is in progress. Recent messages:") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140905 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8221 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123677 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86353 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7835 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5587 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6310 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117231 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41847 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148738 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10007 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42406 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113904 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10024 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8444 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114234 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29043 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7773 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119933 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64731 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10053 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37927 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9784 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73621 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9994 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9845 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -481,8 +481,15 @@ void RenderBox::updateShapeOutsideInfoAfterStyleChange(const RenderStyle& style,
-    // FIXME: A future optimization would do a deep comparison for equality. (bug 100811)
-    if (shapeOutside == oldShapeOutside && shapeMargin == oldShapeMargin && shapeImageThreshold == oldShapeImageThreshold)
+    // Check if shape CSS properties unchanged and that dimension-affecting properties are unchanged
+    // Font, line-height, and writing-mode changes can affect box dimensions even when shape CSS isx identical
```
Typo, `isx` -> `is`.

weinig reviewed (commented):
Added @smfr and @alanbaradlay to the reviewers, as I don't know a ton about this (and if they don't, perhaps they can add the appropriate people).

It's too bad we have to do these super specific checks. I wonder if passing in the `Style::Difference` to the function would be useful? Does the type of difference correspond to any of these new checks?

smfr commented on LayoutTests/fast/shapes/shape-outside-floats/shape-outside-invalidation-after-font-load.html:
```diff
@@ -0,0 +1,76 @@
+    const positionChanged = (Math.abs(newLeft - initialLeft) > 1) || (Math.abs(newTop - initialTop) > 1);
+
+    if (positionChanged) {
+        testPassed(`Content position changed, indicating shape was properly invalidated`);
```
No need for backticks, since there's no substitution.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -481,8 +481,15 @@ void RenderBox::updateShapeOutsideInfoAfterStyleChange(const RenderStyle& style,
+        && oldStyle
+        && oldStyle->fontCascadeEqual(style)
+        && oldStyle->lineHeight() == style.lineHeight()
+        && oldStyle->writingMode() == style.writingMode())
```
There are many other properties that can affect shape-outside, so I'm wondering why we don’t simply rely on the Style::Difference bit. Is there a perf concern around over-invalidation?

nipunshukla21 replied:
Yes, that's correct. I'm not completely sure if it's a valid concern though so I'll run some A/B tests using Style::Difference to see if it regresses.

nipunshukla21 replied:
Using Style::Difference regresses SP3 by 0.82% but progresses MotionMark which doesn't make much sense to me. I've linked the A/B test results in the radar but I think it's best to stick with the current approach.

alanbaradlay replied:
> +    if (oldStyle && Style::difference(style, *oldStyle).result == Style::DifferenceResult::Equal)

This is certainly expensive but also we don't have to do it as the caller (`styleDidChange`) has that diff result already. it just needs to be passed in.

nipunshukla21 replied:
Sorry for missing that. I reran A/B tests in the same task and saw no perf cost to using Style::Difference directly so have updated the patch accordingly. Thanks for the feedback!

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -474,19 +475,15 @@ void RenderBox::updateGridPositionAfterStyleChange(const RenderStyle& style, con
 }
 
-void RenderBox::updateShapeOutsideInfoAfterStyleChange(const RenderStyle& style, const RenderStyle* oldStyle)
+void RenderBox::updateShapeOutsideInfoAfterStyleChange(const RenderStyle& style, const RenderStyle* oldStyle, Style::Difference& diff)
```
Any particular reason diff is passed in by reference here?

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -32,6 +32,7 @@
 #include <WebCore/ScrollSnapOffsetsInfo.h>
 #include <WebCore/ScrollTypes.h>
 #include <WebCore/ShapeOutsideInfo.h>
+#include <WebCore/StyleDifference.h>
```
nit: I am surprised this is needed as this header file already references to StyleDifference.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -474,19 +475,15 @@ void RenderBox::updateGridPositionAfterStyleChange(const RenderStyle& style, con
-
-    // FIXME: A future optimization would do a deep comparison for equality. (bug 100811)
-    if (shapeOutside == oldShapeOutside && shapeMargin == oldShapeMargin && shapeImageThreshold == oldShapeImageThreshold)
+    if (diff == Style::DifferenceResult::Equal)
```
I'd expect it to be safe to check against ::Layout or anything higher than ::RepaintLayer.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -98,6 +98,7 @@
 #include "Settings.h"
 #include "StyleBoxShadow.h"
 #include "StyleComputedStyle+InitialInlines.h"
+#include "StyleDifference.h"
```
nit: I am surprised this is needed as this cpp file already references to StyleDifference.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -474,19 +474,15 @@ void RenderBox::updateGridPositionAfterStyleChange(const RenderStyle& style, con
-
-    // FIXME: A future optimization would do a deep comparison for equality. (bug 100811)
-    if (shapeOutside == oldShapeOutside && shapeMargin == oldShapeMargin && shapeImageThreshold == oldShapeImageThreshold)
+    if (diff > Style::DifferenceResult::RepaintLayer)
```
isn't this backwards? I may be missing something here but I thought we wanna invalidate shapeoutside when we have layouty type of changes.

nipunshukla21 replied:
Yes, you're right -- I think I confused myself. I previously pushed `<= RepaintLayer` based on the idea that it's safe to early return if we're doing paint but we should invalidate if we do layouty stuff as you said. I then changed it to `> RepaintLayer` after re-reading (and misunderstanding) your earlier comment. This is definitely incorrect though so I've changed it back. This is my first time really looking at this code so thanks for keeping up with me here.

alanbaradlay replied:
> if we're doing paint but we should invalidate if we do layouty stuff as you said. I then changed it to > RepaintLayer after re-reading (and misunderstanding) your earlier comment

I should have been clearer. Sorry about that.

nipunshukla21 committed:
REGRESSION (Safari 26.2): shape-outside initially incorrect when used with imported font
https://bugs.webkit.org/show_bug.cgi?id=304019
rdar://166336491

Reviewed by Simon Fraser and Alan Baradlay.

When web fonts load after initial layout, CSS shape properties remain unchanged
but box dimensions can change due to different font metrics. The regression caused
an early return when CSS is identical, preventing shape invalidation.

This issue was fixed by extending the early return condition to check whether
Style::Difference requires a repaint or layout. This ensures shapes are
invalidated when dimension-affecting properties change.

* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-invalidation-after-font-load-expected.txt: Added.
* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-invalidation-after-font-load.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):
(WebCore::RenderBox::updateShapeOutsideInfoAfterStyleChange):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: https://commits.webkit.org/305299@main
